### PR TITLE
ParquetCachedBatchSerializer broadcast AllConfs instead of SQLConf to fix distributed mode

### LINKED ISSUE #2203: [BUG] Spark nightly cache tests fail with -- master flag
**Describe the bug**
When submitting cache tests in the integration tests [here](https://blossom.nvidia.com/sw-gpu-spark-jenkins/job/rapids_integration-dev-github/) with command 

```
spark-submit --master spark://raza-linux-2:7077 --executor-memory 12G --total-executor-cores 6 --conf spark.sql.shuffle.partitions=12 --conf spark.driver.extraClassPath=/home/rjafri/audit-plugin/jars/cudf-0.20-SNAPSHOT-cuda11.jar:/home/rjafri/audit-plugin/jars/rapids-4-spark_2.12-0.6.0-SNAPSHOT.jar:/home/rjafri/audit-plugin/jars/rapids-4-spark-udf-examples_2.12-0.6.0-SNAPSHOT.jar --conf spark.executor.extraClassPath=/home/rjafri/audit-plugin/jars/cudf-0.20-SNAPSHOT-cuda11.jar:/home/rjafri/audit-plugin/jars/rapids-4-spark_2.12-0.6.0-SNAPSHOT.jar:/home/rjafri/audit-plugin/jars/rapids-4-spark-udf-examples_2.12-0.6.0-SNAPSHOT.jar --conf spark.driver.extraJavaOptions=-Duser.timezone=UTC --conf spark.executor.extraJavaOptions=-Duser.timezone=UTC --conf spark.sql.session.timeZone=UTC --conf spark.sql.cache.serializer=com.nvidia.spark.rapids.shims.spark311.ParquetCachedBatchSerializer --jars /home/rjafri/audit-plugin/jars/rapids-4-spark-integration-tests_2.12-0.6.0-SNAPSHOT.jar ./runtests.py -v -rfExXs --std_input_path=/home/rjafri/audit-plugin/integration_tests/src/test/resources/ -k cache_test.py
```

They fail due to an exception. 

``` E                   py4j.protocol.Py4JJavaError: An error occurred while calling o282.collectToPython.
E                   : org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 4.0 failed 4 times, most recent failure: Lost task 0.3 in stage 4.0 (TID 33) (192.168.1.151 executor 0): java.lang.NullPointerException
E                   	at org.apache.spark.sql.internal.SQLConf.setConfWithCheck(SQLConf.scala:3856)
E                   	at org.apache.spark.sql.internal.SQLConf.setConfString(SQLConf.scala:3701)
E                   	at com.nvidia.spark.rapids.shims.spark311.ParquetCachedBatchSerializer$CachedBatchIteratorProducer$InternalRowToCachedBatchIterator.next(ParquetCachedBatchSerializer.scala:1081)
E                   	at com.nvidia.spark.rapids.shims.spark311.ParquetCachedBatchSerializer$CachedBatchIteratorProducer$InternalRowToCachedBatchIterator.next(ParquetCachedBatchSerializer.scala:948)
E                   	at scala.collection.Iterator$$anon$10.next(Iterator.scala:459)
E                   	at org.apache.spark.storage.memory.MemoryStore.putIterator(MemoryStore.scala:222)
E                   	at org.apache.spark.storage.memory.MemoryStore.putIteratorAsValues(MemoryStore.scala:299)
```
Removing the `-- master spark://aaa:bbb` flag from the command makes the script pass

**Steps/Code to reproduce bug**
set WORKSPACE to a local tmp folder and set `URM_URL=https://urm.nvidia.com/artifactory/sw-spark-maven`
run jenkins/spark-tests

**Expected behavior**
All tests should pass

### PR DESCRIPTION
**Pull request by razajafri**
In a distributed cluster, SQLConf is broadcasted as an uninitialized object. This PR broadcasts the conf as a Map and then creates the SQLConf needed in each executor node. 

fixes #2203 

Signed-off-by: Raza Jafri <rjafri@nvidia.com>

### CODE REVIEW

**Comment by razajafri** [comment]
@tgravescs PTAL

**Comment by razajafri** [comment]
build

File: shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala [hunk 1]
```diff
@@ -606,12 +606,18 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
      cbIter: Iterator[CachedBatch],
      cacheAttributes: Seq[Attribute],
      selectedAttributes: Seq[Attribute],
-     sharedHadoopConf: Configuration,
-     sharedConf: SQLConf) {
+     sharedHadoopConf: Broadcast[SerializableConfiguration],
+     sharedConf: Broadcast[Map[String, String]]) {
 
+    val conf = {
```
Comment (tgravescs) [inline shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala:612]: write utility function so we don't have to duplicate code

**Review by tgravescs — COMMENTED** [review]
seems fine a the high level to fix the issue with broadcast sqlconf being empty, I didn't have time to look at the details what its being used for, but seems safe if setting all of them.

**Comment by razajafri** [comment]
build

File: tests-spark310+/src/test/scala/com/nvidia/spark/rapids/Spark310ParquetWriterSuite.scala [hunk 2]
```diff
@@ -167,8 +169,10 @@ class Spark310ParquetWriterSuite extends SparkQueryCompareTestSuite {
       }
     }
     val ser = new ParquetCachedBatchSerializer
+
     val producer = new ser.CachedBatchIteratorProducer[ColumnarBatch](cbIter, schema,
-      new Configuration(true), new SQLConf)
+      SparkContext.getOrCreate().broadcast(new SerializableConfiguration(new Configuration(true))),
```
Comment (gerashegalov) [inline tests-spark310+/src/test/scala/com/nvidia/spark/rapids/Spark310ParquetWriterSuite.scala:174]: curios why `withGpu/CpuSparkSession { sparkSession => sparkSession.sparkContext ...` is not used here

File: shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala [hunk 3]
```diff
@@ -929,8 +936,11 @@ class ParquetCachedBatchSerializer extends CachedBatchSerializer with Arm {
   private[rapids] class CachedBatchIteratorProducer[T](
      iter: Iterator[T],
      cachedAttributes: Seq[Attribute],
-     sharedHadoopConf: Configuration,
-     sharedConf: SQLConf) {
+     sharedHadoopConf: Broadcast[SerializableConfiguration],
```
[the remark this answers is not in the document]
Comment (razajafri) [reply shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala:939]: This is a new HadoopConf that was created, without any default values, with custom values that were needed by this class to write a Parquet file in memory. 

Now that we know its not that huge, what's the advantage of not passing this conf but a Map other than consistency?

File: tests-spark310+/src/test/scala/com/nvidia/spark/rapids/Spark310ParquetWriterSuite.scala (hunk 2 above)
Comment (razajafri) [reply tests-spark310+/src/test/scala/com/nvidia/spark/rapids/Spark310ParquetWriterSuite.scala:174]: No reason other than I forgot to use them as I haven't written a scala test in our plugin in a long time.

**Comment by razajafri** [comment]
build

File: shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (gerashegalov) [reply shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala:939]: that's right, for consistency.

[the remark this answers is not in the document]
Comment (razajafri) [reply shims/spark311/src/main/scala/com/nvidia/spark/rapids/shims/spark311/ParquetCachedBatchSerializer.scala:939]: You just made me realize something. Let me push another update, we may not need to pass the conf at all.

File: tests-spark310+/src/test/scala/com/nvidia/spark/rapids/Spark310ParquetWriterSuite.scala [hunk 4]
```diff
@@ -167,8 +169,11 @@ class Spark310ParquetWriterSuite extends SparkQueryCompareTestSuite {
       }
     }
     val ser = new ParquetCachedBatchSerializer
+
     val producer = new ser.CachedBatchIteratorProducer[ColumnarBatch](cbIter, schema,
-      new Configuration(true), new SQLConf)
+      withCpuSparkSession(spark =>
```
Comment (gerashegalov) [inline tests-spark310+/src/test/scala/com/nvidia/spark/rapids/Spark310ParquetWriterSuite.scala:174]: nit: may not matter here but withGpuSparkSession is probably better

**Review by gerashegalov — APPROVED** [review]
LGTM, but we need an issue to clarify whether and what is the minimum config keys we need to broadcast manually.

**Pull request merged**